Added the git clone command and all associated functionality and tests

Tyler-g-hudson opened the pull request:
This PR includes all of the functionality to clone a git repository into a docker image.

Tyler-g-hudson committed:
Added the git clone command and all associated functionality and tests

gmgunter commented on docker_cli/_docker_git.py:
```diff
@@ -0,0 +1,117 @@
+    folder_name: str = "repo",
+) -> Tuple[str, str]:
+    """
+    Returns a dockerfile-formatted string with instructions to clone a git repository.
```
```suggestion
    Returns a Dockerfile-formatted string with instructions to clone a Git repository.
```

gmgunter commented on docker_cli/_docker_git.py:
```diff
@@ -0,0 +1,117 @@
+    git_url : _type_, optional
+        The URL to get the git repository from. Defaults to "https://github.com/".
+    folder_name : str, optional
+        The name of the folder to store the repository in. Defaults to "repo".
```
Let's make this a path-like object representing the absolute path in the container filesystem where the git repo should be cloned, rather than assuming that the user wants this directory to live just under the root directory

gmgunter commented on docker_cli/_docker_git.py:
```diff
@@ -0,0 +1,117 @@
+        + "\n"
+    )
+
+    # Activate Micromamba
```
```suggestion
    # Switch user to `MAMBA_USER`
```

gmgunter commented on docker_cli/_docker_git.py:
```diff
@@ -0,0 +1,117 @@
+    repo_branch : str, optional
+        The name of the branch to checkout. Defaults to "". Currently disabled.
+    git_url : _type_, optional
+        The URL to get the git repository from. Defaults to "https://github.com/".
```
Let's combine `git_repo` and `git_url` into a single argument for simplicity

gmgunter commented on docker_cli/_docker_git.py:
```diff
@@ -0,0 +1,117 @@
+    body += (
+        dedent(
+            f"""
+        ADD {git_url}/{git_repo}.git /{folder_name}/
```
Since `ADD`ing a git repo is currently an experimental feature of docker, let's avoid that potential complexity by just fetching a particular release instead. We could have the user simply provide the URL of the git archive to fetch here instead.

```suggestion
        ADD https://github.com/isce-framework/isce3/archive/refs/tags/v0.14.0.tar.gz /{folder_name}/
```

gmgunter commented on docker_cli/_docker_git.py:
```diff
@@ -0,0 +1,117 @@
+    return header, body
+
+
+def git_clone_command(
```
I think we prefer to remove this for now and stick to the use case above. But it may be a nice future addition in a later PR

gmgunter commented on docker_cli/_docker_git.py:
```diff
@@ -0,0 +1,117 @@
+    header = "# syntax=docker/dockerfile:1-labs"
+
+    # Return the generated body plus a header
+    return header, body
```
I think I want to start moving away from returning a "header" and "body" in methods like this. Instead, let's have the user provide the base image (or base tag) as an input argument so that we can fully form the Dockerfile and return that.

gmgunter commented on docker_cli/cli/build_commands.py:
```diff
@@ -0,0 +1,60 @@
+        help="The name of the GitHub repository to be installed. "
+        'Default: "isce-framework/isce3"',
+    )
+    # This argument currently disabled as this code does not presently support git
+    # branches. For consideration on this topic, see _docker_git.py in the parent folder
+    """clone_params.add_argument(
+        "--branch", type=str, metavar="REPO_BRANCH", default="",
+        help="The name of the branch to checkout. Defaults to \"\"."
+    )"""
```
Let's remove this since we don't plan to support git branches right now

```suggestion
```

gmgunter commented on docker_cli/cli/build_commands.py:
```diff
@@ -0,0 +1,60 @@
+        "--repo",
+        type=str,
+        metavar="GIT_REPO",
+        default="isce-framework/isce3",
```
Given our changes to the lower-level function, let's remove this default value and make this parameter required

```suggestion
```

gmgunter commented on docker_cli/cli/build_commands.py:
```diff
@@ -0,0 +1,60 @@
+    Augment an argument parser with build commands.
+
+    Parameters
+    -------
```
```suggestion
    -----------
```

gmgunter commented on docker_cli/cli/build_commands.py:
```diff
@@ -0,0 +1,60 @@
+
+def init_build_parsers(subparsers: argparse._SubParsersAction) -> None:
+    """
+    Augment an argument parser with build commands.
```
Can we clarify what we mean by "build commands". Ideally, an arbitrary person on the isce3 development team could read this docstring and get a reasonable understanding of what's going on.

gmgunter commented on docker_cli/commands.py:
```diff
@@ -1,13 +1,66 @@
+    img_tag = tag if tag.startswith(prefix) else f"{prefix}-{tag}"
+
+    header, body = git_clone_dockerfile(
+        git_repo=repo, repo_branch=branch, folder_name=repo_name
```
Choosing the output `folder_name` is going to be trickier with the changes we've discussed offline. Maybe it's best to have the user provide this directory separately. You could default the name to something like `/src`

gmgunter commented on docker_cli/commands.py:
```diff
@@ -91,8 +144,8 @@ def make_lockfile(
     tag : str
-        The tag or ID of the image.
-    file : Union[str, os.PathLike[str]]
+        The tag of the image
```
```suggestion
        The tag of the image.
```

gmgunter commented on test/test_docker_git.py:
```diff
@@ -0,0 +1,11 @@
+    header, body = git_clone_dockerfile(git_repo="abc/def", repo_branch="ghi")
+
+    dockerfile = f"{header}\n\nFROM %TEST_BASE%\n\n{body}"
+    rough_dockerfile_validity_check(dockerfile=dockerfile)
```
Let's improve this test to make sure that we're checking:

(A) the git repo was actually cloned inside the image
(B) that the current working directory is now the cloned directory

Here's a simple git repo that you could use in your test: https://github.com/octocat/Hello-World

gmgunter reviewed (commented):
Thanks! Here's a few notes from our discussion this afternoon

Tyler-g-hudson committed:
Modified the method of retrieval to extract a git archive instead of cloning the repo.

~ Substantial modifications to `git_clone_dockerfile` (now renamed `git_extract_dockerfile`.)
~ Renamed the command to `get-archive`
~ Modified the interface accordingly

Tyler-g-hudson committed:
CLI changes

+ Added a deeper description of the build commands in `init_build_parsers`
~ Made the `--archive-url command` required
~ Changed `clone` to `archive` for accuracy.

Tyler-g-hudson committed:
Added a better test, a new pytest mark, and added tar as a requirement.

+ New init images now check for the `tar` command and add it if it isn't present.
+ Added a `git` mark for pytests.
+ Added a test that acquires a git archive and tests for files on it.

gmgunter commented on docker_cli/_docker_git.py:
```diff
@@ -0,0 +1,77 @@
+    base: str,
+    archive_url: str,
+    url_reader: URLReader,
+    folder_path: os.PathLike[str] = Path("repo"),
```
```suggestion
    folder_path: str | os.PathLike[str] = Path("repo"),
```

(this requires `from __future__ import annotations` at the top)

gmgunter commented on docker_cli/_docker_git.py:
```diff
@@ -0,0 +1,77 @@
+        The URL of the Git archive. Must be a `tar.gz` file.
+    url_reader : URLReader
+        The URL reader program to fetch the archive with.
+    folder_name : path-like, optional
```
Let's rename "folder" to "directory", which I think is the more widely used term.

```suggestion
    directory : path-like, optional
```

gmgunter commented on docker_cli/_docker_git.py:
```diff
@@ -0,0 +1,77 @@
+    dockerfile : str
+        The generated Dockerfile.
+    """
+    folder_path_str = str(folder_path)
```
```suggestion
    folder_path_str = os.fspath(folder_path)
```

gmgunter commented on docker_cli/_docker_git.py:
```diff
@@ -0,0 +1,77 @@
+    """
+        ).strip()
+        + "\n"
+    )
```
The indentation is kinda weird here. What's nice about using `dedent()` and `strip()` is that you can make the indentation of the string body here match the indentation of the surrounding code.

Here's how I would probably indent this code:

```suggestion
    dockerfile = (
        dedent(f"""
            FROM {base}
    
            USER root
    
            RUN mkdir -p {folder_path_str}
            RUN chown -R $MAMBA_USER_ID:$MAMBA_USER_GID {folder_path_str}
            RUN chmod -R 755 {folder_path_str}
        """).strip()
        + "\n"
    )
```

It looks like that's pretty close to [what black would do in `--preview` mode](https://black.readthedocs.io/en/stable/the_black_code_style/future_style.html#improved-multiline-string-handling)

Tyler-g-hudson replied:
My version of `black` doesn't seem to like this - in particular, it doesn't like `dedent(f"""` and `""").strip` in which the multi-line string is begun or ended on the same line as the opening or closing of the `dedent` function.

I've pushed a version that `black` doesn't complain about on my end, and looks a bit better than my prior code. LMK what you think.

gmgunter replied:
Cool, thanks. I like your updated formatting. Let's do something similar here:

https://github.com/isce3-testing/image-sandbox/blob/1f6658743964d439999f611affc73e85a44dff2a/docker_cli/_docker_git.py#L65-L75

gmgunter commented on docker_cli/cli/build_commands.py:
```diff
@@ -0,0 +1,67 @@
+    The build commands are the group of commands to be completed after the CUDA and
+    conda environments have been installed to the image, with the purpose of acquiring
+    and building the ISCE3 repository and any further repositories.
+    These commands consist of the "get-archive" command, and more are being added.
```
Nice :+1:

gmgunter commented on docker_cli/cli/build_commands.py:
```diff
@@ -0,0 +1,67 @@
+    archive_params.add_argument(
+        "--folder-path",
+        type=Path,
+        metavar="FOLDER_PATH",
```
```suggestion
        "--directory",
        type=Path,
        metavar="DIRECTORY",
```

gmgunter commented on docker_cli/commands.py:
```diff
@@ -1,11 +1,69 @@
+    url_reader: URLReader | None = None,
+):
+    """
+    Builds a docker image containing the requested Git repository.
```
```suggestion
    Builds a docker image containing the requested Git archive.
```

gmgunter commented on test/test_docker_git.py:
```diff
@@ -0,0 +1,55 @@
+    # in the right location.
+    image.run("test -f /src/2015-04-12-test-post-last-year.md")
+    image.run("test -f /src/2016-02-24-first-post.md")
+    image.run("test -f /src/2016-02-26-sample-post-jekyll.md")
```
This test looks great, thanks

gmgunter reviewed (commented):
Good stuff. Few more minor nitpicks but overall looks good to me.

Tyler-g-hudson committed:
`folder` -> `directory`, some documentation changes.

gmgunter reviewed (approved):
LGTM, nice work!

Tyler-g-hudson committed:
Updated indentation on git archive acquisition command
